NIFI-3785.1: Added feature to move controller services between process …

Freedom9339 opened the pull request:
…groups with the new UI

<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

[NIFI-3785](https://issues.apache.org/jira/browse/NIFI-3785)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [X] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [X] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [X] Pull Request based on current revision of the `main` branch
- [X] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [X] Build completed using `mvn clean install -P contrib-check`
  - [X] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

Freedom9339 commented:
@markap14  @exceptionfactory I've resubmitted this PR with the changes done to the new UI. The backend is the same as it was in the closed PR(https://github.com/apache/nifi/pull/7734), but the UI changes were done against the new UI.

Freedom9339 commented:
@markobean

exceptionfactory commented:
Thanks for putting in the work to refactor this for the new UI @Freedom9339.

Tagging @mcgilman and @rfellows for potential review.

mcgilman commented:
Thanks for the mention @exceptionfactory! I'll have a look...

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+                }
+            }
+
+            processGroupflow.flow.processGroups.forEach((child:any) => {
+                if (child.permissions.canRead && child.permissions.canWrite) {
+                    let option: SelectOption = {
+                        text: child.component.name,
+                        value: child.component.id
+                    }
+                    processGroups.push(option);
+                }
+            });
```
We should only allow the user to choose a child component if the reduction in Controller Service scope will be allowed. By moving to a child Group if there are any components referencing that Controller Service in the current Group won't be able to.

Freedom9339 replied:
The reason I had not limited the process group list is because if the user does try and move a controller service that has referencing components that outside of the new scope, the API returns an error listing the components that have a conflict. If I filter out the list, the user won't know why it's not in the drop down.

I can change this to only display viable process groups if that is what is wanted though.

mcgilman replied:
Thoughts on adding it to the listing but disabling it and adding a description as to why it cannot be selected? The `SelectOption` already supports `disabled` and `description` field so it should just be a matter of conditionally populating the fields. Though you may need to also account for them in your template.

Freedom9339 replied:
I've implemented this change. The dropdown menu will still list all parent and child process groups, but any process group with a conflict will not be disabled with an ! icon and a hover tooltip explaining the conflict. 

I will note this does require the use of 2 api calls. I also had to add an optional parameter to an api definition.

Freedom9339 replied:
Actually, I thought of something about 2 seconds after I pushed. Found a way to avoid needed an api call. So now we only have 1 api call. Tested and pushed.

mcgilman replied:
Shouldn't we be able to determine this without any additional API calls? Within the Controller Service we have a list of referencing components which contain their Parent Process Group ID. We also have the breadcrumbs which contain the current Process Group and all of the ancestors Process Groups.

If the referencing component's Parent Group ID is one of the breadcrumbs than it would mean that the Controller Service would be out of scope if referencing component were moved to a child Process Group.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+            processGroups: new FormControl("Process Group", Validators.required)
+        });
+
+        this.store.dispatch(
+            setControllerService({
+                request: {
+                    controllerService: request.controllerService
+                }
+            })
+        );
```
I don't think this component needs to store the subject Controller Service in a store.

Freedom9339 replied:
Removed this.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+import { NifiSpinnerDirective } from '../../spinner/nifi-spinner.directive';
+import { TextTip } from '../../tooltips/text-tip/text-tip.component';
+import { NifiTooltipDirective } from '../../tooltips/nifi-tooltip.directive';
+import { ControllerServiceState } from '../../../../state/contoller-service-state';
```
This component shouldn't be referencing `ControllerServiceState`.

Freedom9339 replied:
Removed this.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+
+    constructor(
+        @Inject(MAT_DIALOG_DATA) public request: MoveControllerServiceDialogRequest,
+        private store: Store<ControllerServiceState>,
```
```suggestion
        private store: Store<NiFiState>,
```

Freedom9339 replied:
Made this change. Now using NiFiState store.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+    protected readonly TextTip = TextTip;
+    protected controllerServiceActionProcessGroups: SelectOption[] = [];
+
+    controllerService$ = this.store.select(selectControllerService);
```
I don't think this component needs to select anything from a store. The Controller Service in question is supplied by the `request` constructor argument.

Freedom9339 replied:
removed this.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/state/shared/index.ts:
```diff
@@ -132,6 +132,12 @@ export interface EditControllerServiceDialogRequest {
     history?: ComponentHistory;
 }
 
+export interface MoveControllerServiceDialogRequest {
+    id: string;
+    controllerService: ControllerServiceEntity;
+    processGroupFlow: any;
+}
```
Like the new `Move` component, I think this dialog request should be moved into the `flow-designer` page. It's not applicable outside of that page.

mcgilman replied:
`processGroupFlow` shouldn't need to be `any` for this dialog request. Admittedly there is some instances of `any` based on the underlying data model the back end uses but in this case there already is a defined type for this field that should be referenced.

Freedom9339 replied:
Moved this to flow-designer. Also, changed to use the ProcessGroupFlow model.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -635,6 +636,82 @@ export class ControllerServicesEffects {
+                concatLatestFrom(() => this.store.select(selectCurrentProcessGroupId)),
+                tap(([request, currentProcessGroupId]) => {
+                    const serviceId: string = request.id;
+                    request.processGroupFlow = this.flowService.getFlow(currentProcessGroupId);
```
The state for the `flow-designer` page is already loaded with the current process group. We should be able to simply use that instead of making another request here.

Freedom9339 replied:
The current state did not store the entire flow. I did change this and I was able to use it here without an api call. However, with the implementation of disabled drop down menu process groups, I did need to 2 other api calls here.

mcgilman replied:
Ah sorry. Yes you are correct. I thought we had stashed the entire `processGroupFlow` in the service state as well but I see that only saved the breadcrumbs and parameter context details. Now we're saving the entire `processGroupFlow` next to the breadcrumbs and parameter context details.

Do we only need the `processGroupFlow` in order to get the child Process Groups for the Move dialog? If so, what are your thoughts only saving them in the store. Often times the flow can be very large and in this page we won't need most of those details.

Let me know if I missed something and we do need additional details.

Freedom9339 replied:
Thank you for the idea, I hadn't thought of that. I removed the storing of the whole process group flow, and just saved an array with names and IDs.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.spec.ts:
```diff
@@ -0,0 +1,363 @@
+import { MAT_DIALOG_DATA, MatDialogModule, MatDialogRef } from '@angular/material/dialog';
+import { ComponentType, MoveControllerServiceDialogRequest } from '../../../../state/shared';
+
+describe('MoveControllerService', () => {
```
This test currently does not run.

> Summary of all failing tests
>  FAIL  src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.spec.ts
>   ● MoveControllerService › should create

Freedom9339 replied:
Fixed the test. It now passes.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -635,6 +636,82 @@ export class ControllerServicesEffects {
+                    const serviceId: string = request.id;
+                    request.processGroupFlow = this.flowService.getFlow(currentProcessGroupId);
+                    const moveDialogReference = this.dialog.open(MoveControllerService, {
+                        ...XL_DIALOG,
```
Changed this to a LARGE_DIALOG.

mcgilman replied:
The dialog size looks great. Can we update it so the content shares the available space like other dialogs (for instance Edit Controller Service)?

![Screenshot 2024-07-02 at 5 18 11 PM](https://github.com/apache/nifi/assets/123395/1bf341cd-ad19-4dea-8955-e5eeecadfafd)

Freedom9339 replied:
I fixed this. It now extends all the way.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -635,6 +636,82 @@ export class ControllerServicesEffects {
+                    request.controllerService.parentGroupId,
+                    'controller-services',
+                    request.controllerService.id
+                ]);
```
Removed post navigation.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.reducer.ts:
```diff
@@ -99,6 +100,15 @@ export const controllerServicesReducer = createReducer(
+            }
+            draftState.saving = false;
+        });
+    }),
```
Changed the reducer to remove the controller service.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/controller-service-table/controller-service-table.component.html:
```diff
@@ -173,6 +173,12 @@
+                                        <i class="fa fa-arrows primary-color mr-2"></i>
+                                        Move
+                                    </button>
+                                }
```
Made the move option not show up in  management controller services.

mcgilman replied:
Thanks for the update here. The new `Input()` supports function that returns whether the service can be moved. However, the implements are not service specific and always just return `true` or `false`. Can this `Input()` be simplified into a boolean?

`@Input() canMove: boolean = false;`

Freedom9339 replied:
Changed this to just be a boolean.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,82 @@
+<ng-template #stepError>
+    <div class="fa fa-times warn-color"></div>
+</ng-template>
+<ng-template #stepNotStarted><div class="w-3.5"></div></ng-template>
```
Removed these.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+    @ViewChild('stepComplete') stepComplete!: TemplateRef<any>;
+    @ViewChild('stepError') stepError!: TemplateRef<any>;
+    @ViewChild('stepInProgress') stepInProgress!: TemplateRef<any>;
+    @ViewChild('stepNotStarted') stepNotStarted!: TemplateRef<any>;
```
Removed these.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+    ],
+    styleUrls: ['./move-controller-service.component.scss']
+})
+export class MoveControllerService extends CloseOnEscapeDialog {
```
Moved the move component to flow-designer.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+import { ControllerServiceState } from '../../../../state/contoller-service-state';
+import { Store } from '@ngrx/store';
+import {
+    setControllerService,
```
Changed the store to NiFiState.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/ui/common/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,158 @@
+
+    cancelClicked(): void {
+        this.store.dispatch(stopPollingControllerService());
+    }
```
Removed this.

Freedom9339 commented:
@mcgilman Just checking in. Any updates on this?

mcgilman commented:
> @mcgilman Just checking in. Any updates on this?

Thanks for the ping. Sorry for the delay, I'll try to get some more eyes on this PR this week.

Freedom9339 commented:
@mcgilman Sorry the pinging again. Just want to make sure this doesn't fall through the cracks.

mcgilman commented:
> @mcgilman Sorry the pinging again. Just want to make sure this doesn't fall through the cracks.

I just pulled down the latest. After rebasing against current `main` there were build failures due to changes that have landed after your PR was created. So I tried to run the PR as-is without rebasing and I'm still seeing issues. When running through the dev server it fails to serve with

> Application bundle generation failed. [13.461 seconds]
> 
> ✘ [ERROR] Undefined function.
>     ╷
> 109 │ $surface-light-palette: mat.define-palette($surface-palette, 600, 100, 900);
>     │                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>     ╵
>   libs/shared/src/assets/themes/supplemental.scss 109:25  @import
>   apps/nifi/src/styles.scss 62:9                          root stylesheet [plugin angular-sass]
> 
>     angular:styles/global:styles:1:8:
>       1 │ @import 'apps/nifi/src/styles.scss';

I then tried running with the built application but I'm seeing this error in dev tools when attempting to open the `MoveControllerService` Component.

```
TypeError: g is not iterable
    at chunk-VIZEOFFB.js:70:66450
    at Array.forEach (<anonymous>)
    at o.loadChildOptions (chunk-VIZEOFFB.js:70:66333)
    at new o (chunk-VIZEOFFB.js:70:65423)
    at o.ɵfac [as factory] (chunk-VIZEOFFB.js:70:67428)
    at hn (chunk-ZKVQE3GD.js:7:23746)
    at TC (chunk-ZKVQE3GD.js:7:63202)
    at mn.create (chunk-ZKVQE3GD.js:7:61795)
    at Vg.createComponent (chunk-ZKVQE3GD.js:7:64954)
    at t.attachComponentPortal (chunk-6XYXU2WR.js:1:110675)
```
Once the PR is updated I should be able to get some more eyes on it quickly.

Freedom9339 commented:
@mcgilman I fixed the issue caused by rebasing and pushed. Thank You!

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,232 @@
+                value: child.value
+            };
+
+            const root = this.getProcessGroupById(currentProcessGroupEntity.component, child.value ?? '');
```
`getProcessGroupById` will return null for a couple different scenarios. When this happens calling `processGroupContainsComponent` will error with the following. The unhandled error leaves the UI in a weird state because the overlay is active but the dialog isn't shown.

```
TypeError: Cannot read properties of null (reading 'contents')
    at _MoveControllerService.processGroupContainsComponent (move-controller-service.component.ts:184:26)
    at move-controller-service.component.ts:166:27
    at Array.forEach (<anonymous>)
    at _MoveControllerService.loadChildOptions (move-controller-service.component.ts:157:34)
    at new _MoveControllerService (move-controller-service.component.ts:99:14)
    at NodeInjectorFactory.MoveControllerService_Factory [as factory] (move-controller-service.component.ts:231:5)
    at getNodeInjectable (core.mjs:6060:44)
    at createRootComponent (core.mjs:17231:35)
    at ComponentFactory.create (core.mjs:17091:29)
    at ViewContainerRef2.createComponent (core.mjs:17498:47) 
```

mcgilman replied:
It appears this was happening when `childProcessGroupOptions` contained more than one option. The recursive logic in `getProcessGroupById` (and other methods with similar structure) isn't correct. When there are multiple children the following code

```
                    for (const pg of root.contents.processGroups) {
                        return this.getProcessGroupById(pg, processGroupId);
                    }
```

will return the result from the first child and not attempt to process any subsequent children.

Freedom9339 replied:
I've fixed the recursion logic.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -63,7 +66,8 @@ import {
 import { BackNavigation } from '../../../../state/navigation';
-import { NiFiCommon, Storage } from '@nifi/shared';
+import { NiFiCommon, Storage, SelectOption, ComponentType, LARGE_DIALOG, SMALL_DIALOG, XL_DIALOG } from '@nifi/shared';
+import { ComponentEntity } from './../flow/index';
```
```suggestion
import { ComponentEntity } from '../flow';
```

Freedom9339 replied:
Done.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/index.ts:
```diff
@@ -15,8 +15,10 @@
+import { SelectOption } from '@nifi/shared';
 import { ControllerServiceEntity, ParameterContextReferenceEntity } from '../../../../state/shared';
 import { BreadcrumbEntity } from '../shared';
+import { Revision } from './../../../../state/shared/index';
```
```suggestion
import { Revision } from '../../../../state/shared';
```

Freedom9339 replied:
Done.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,232 @@
+        );
+
+        const processGroups: SelectOption[] = [];
+        if (request.breadcrumb != undefined) {
```
Under what conditions is `request.breadcrumb` `undefined`? I think the store has an initial value but even if that isn't the case we could update the `concatLatestFrom` with the following:

`this.store.select(selectBreadcrumb).pipe(isDefinedAndNotNull())`

Freedom9339 replied:
I was trying to only use a single request model, but I've created 2 and removed the need to declare it as `breadcrumb?`

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -651,6 +666,101 @@ export class ControllerServicesEffects {
+                        )
+                ),
+                tap((request) => {
+                    const clone = Object.assign({}, request.request);
+                    clone.processGroupEntity = request.processGroupEntity;
+                    clone.childProcessGroupOptions = request.childProcessGroupOptions;
+                    clone.parentControllerServices = request.controllerServices;
+                    clone.breadcrumb = request.breadcrumb;
```
Why is a clone being created here?

Freedom9339 replied:
Same as above. Removed need to clone.

mcgilman reviewed (commented):
Thanks for the rebase! I've left some additional feedback below.

Freedom9339 commented:
@mcgilman Sorry, I just realized my latest commit didn't go through. It's there now.

Freedom9339 commented:
@mcgilman Any update on this? I just did a fresh rebase.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,82 @@
+                <div>
+                    <div>Service</div>
+                    <div
+                        class="accent-color font-medium overflow-ellipsis overflow-hidden whitespace-nowrap"
```
```suggestion
                        class="tertiary-color font-medium truncate"
```

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,82 @@
+<form class="controller-service-move-form" [formGroup]="moveControllerServiceForm">
+    <mat-dialog-content>
+        <div class="py-4 flex gap-x-3">
+            <div class="flex basis-2/3 flex-col gap-y-4 pr-4 overflow-hidden">
```
```suggestion
            <div class="flex basis-1/2 flex-col gap-y-4 pr-4 overflow-hidden">
```

Thoughts on splitting the dialog space, half for the Group selection and half for the References?

Freedom9339 replied:
I agree it looks better. Done.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,82 @@
+                    </mat-form-field>
+                </div>
+            </div>
+            <div class="flex basis-1/3 flex-col">
```
```suggestion
            <div class="flex basis-1/2 flex-col">
```

mcgilman reviewed (commented):
Thanks for the updates @Freedom9339! I've left a few more comments below.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    getProcessGroupById(root: any, processGroupId: string): any {
+        console.log('a');
```
It was. I've removed it.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,82 @@
+    </mat-dialog-content>
+    <mat-dialog-actions align="end">
+        <button mat-button mat-dialog-close>Cancel</button>
+        <button type="button" color="primary" (click)="submitForm()" mat-button>Move</button>
```
The button is now disabled when there no valid options.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    submitForm() {
+        if (this.moveControllerServiceForm.get('processGroups')?.value != 'Process Group') {
```
Removed.

Freedom9339 commented:
@mcgilman Thank you for the review. I've made the requested changes. Thank You.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+
+        // build the form
+        this.moveControllerServiceForm = this.formBuilder.group({
+            processGroups: new FormControl('Process Group', Validators.required)
```
`Process Group` is not a valid value for this field. We could initialize the value to `null` but we're setting the value below. We could instead determine the value before creating the form and then just use the correct value when creating the `FormControl`.

Freedom9339 replied:
Not sure what you mean by this. "Process Group" is what is shown whenever there is no process group selected. The only time this happens is if there are no valid process groups available.

mcgilman replied:
The first argument to the FormControl is the value. The string `Process Group` is not a valid value. What you see in the control when no value is select comes from the `mat-label` in the `mat-form-field`. If you want something else than the `mat-label` value you can specify a `placeholder` on the `mat-select`.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+            const parentBreadcrumb = breadcrumb.parentBreadcrumb;
+            if (parentBreadcrumb.permissions.canRead && parentBreadcrumb.permissions.canWrite) {
+                const option: SelectOption = {
+                    text: parentBreadcrumb.breadcrumb.name + ' (Parent)',
```
Please convert to a template string.

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+                            controllerId != null &&
+                            !parentControllerServices.some((service) => service.id == controllerId)
+                        ) {
+                            errorMsg += '[' + descriptors[descriptor].name + ']';
```
Please convert to a template string if possible.

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+                }
+
+                if (errorMsg != '') {
+                    option.description =
+                        'The following properties reference controller services that would be out of scope for this ' +
+                        'process group: ' +
+                        errorMsg;
```
Please convert to a template string.

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+            } else {
+                for (const component of referencingComponents) {
+                    if (!this.processGroupContainsComponent(root, component.component.groupId)) {
+                        errorMsg += '[' + component.component.name + ']';
```
Please convert to a template string if possible.

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+                }
+
+                if (errorMsg != '') {
+                    option.description =
+                        'The following components would be out of scope for this process group: ' + errorMsg;
```
Please convert to a template string.

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    getProcessGroupById(root: any, processGroupId: string): any {
+        if (root != undefined) {
```
```suggestion
        if (root) {
```

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    processGroupContainsComponent(processGroup: any, groupId: string): boolean {
+        if (processGroup.contents != undefined) {
```
```suggestion
        if (processGroup.contents) {
```

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+                let errorMsg = '';
+                const descriptors = this.controllerService.component.descriptors;
+                for (const descriptor in descriptors) {
+                    if (descriptors[descriptor].identifiesControllerService != undefined) {
```
```suggestion
                    if (descriptors[descriptor].identifiesControllerService) {
```

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+            const root = this.getProcessGroupById(currentProcessGroupEntity.component, child.value ?? '');
+            let errorMsg = '';
+
+            if (root == null) {
```
It may make this more readable  if we swap the `if/else` blocks and update the conditional to:

```suggestion
            if (root) {
```

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+            if (root.id == processGroupId) {
+                return root;
+            } else {
+                if (root.contents != undefined) {
```
```suggestion
                if (root.contents) {
```

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+                if (root.contents != undefined) {
+                    for (const pg of root.contents.processGroups) {
+                        const result = this.getProcessGroupById(pg, processGroupId);
+                        if (result != null) {
```
```suggestion
                        if (result) {
```

Freedom9339 replied:
No longer used.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/service/flow.service.ts:
```diff
@@ -118,6 +118,11 @@ export class FlowService implements PropertyDescriptorRetriever {
         return this.httpClient.get(`${FlowService.API}/process-groups/${id}`);
     }
 
+    getProcessGroupWithContent(id: string): Observable<any> {
+        const params = new HttpParams().set('includeContent', true);
+        return this.httpClient.get(`${FlowService.API}/process-groups/${id}`, { params: params });
+    }
```
This change is necessary for this feature to ensure that when moving a Controller Service it remains in scope for all components currently referencing it. However, I have some concerns about fetching the entire flow from the current Process Group when the Move dialog is opened. I wonder if we should consider a purpose built endpoint for does exactly what we need instead of returning everything. We've certainly seen some large flows over the years where I worry this could be problematic.

CC @exceptionfactory

Freedom9339 replied:
I've created a new API endpoint, move-options, to get the information needed for the process groups. This shifts all the logic to the API, and greatly simplifies the UI form.

mcgilman replied:
I agree. This refactoring has made the UI much simpler.

mcgilman reviewed (commented):
@Freedom9339 Thanks again for the updates! It's looking good. I've left a couple comments and I've noted a number of nit-picky coding style things to better follow some of the conventions currently practiced.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,247 @@
+    ],
+    styleUrls: ['./move-controller-service.component.scss']
+})
+export class MoveControllerService extends CloseOnEscapeDialog {
```
I've moved this to ui/controller-service

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,80 @@
+    </mat-dialog-content>
+    <mat-dialog-actions align="end">
+        <button mat-button mat-dialog-close>Cancel</button>
+        <button type="button" color="primary" (click)="submitForm()" [disabled]="disableSubmit" mat-button>Move</button>
```
I've changed this to use the state of the form.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/controller-service/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,87 @@
+                </div>
+                <div>
+                    <mat-form-field>
+                        <mat-label>Process Group:</mat-label>
```
```suggestion
                        <mat-label>Process Group</mat-label>
```
Other `mat-labels` in the UI don't include a colon.

Freedom9339 replied:
Removed Colon

mcgilman reviewed (commented):
Thanks for the updates @Freedom9339. I've left a few comments below. I didn't call out a specific line but the new `move-options` endpoint should not be providing options and exposing details for parent or child Process Groups that the end user does not have permissions to. Please ensure your trying out all these scenarios when your PR changes.

Thanks!

mcgilman commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ControllerServiceResource.java:
```diff
@@ -683,6 +686,209 @@ public Response updateControllerService(
+        controllerServiceDTO.getReferencingComponents().forEach(e -> {
+            if (processGroup.getProcessGroup().findProcessor(e.getId()) == null
+                    && processGroup.getProcessGroup().findControllerService(e.getId(), true, false) == null) {
+                conflictingComponents.add("[" + e.getComponent().getName() + "]");
```
If I take away permissions from a component that references the Service I'm trying to remove this line generates a `NullPointerException`

```
2024-11-18 16:04:51,256 ERROR [NiFi Web Server-76] o.a.nifi.web.api.config.ThrowableMapper An unexpected error has occurred: java.lang.NullPointerException: Cannot invoke "org.apache.nifi.web.api.dto.ControllerServiceReferencingComponentDTO.getName()" because the return value of "org.apache.nifi.web.api.entity.ControllerServiceReferencingComponentEntity.getComponent()" is null. Returning Internal Server Error response.
java.lang.NullPointerException: Cannot invoke "org.apache.nifi.web.api.dto.ControllerServiceReferencingComponentDTO.getName()" because the return value of "org.apache.nifi.web.api.entity.ControllerServiceReferencingComponentEntity.getComponent()" is null
    at org.apache.nifi.web.api.ControllerServiceResource.lambda$generateProcessGroupOption$18(ControllerServiceResource.java:751)
```

Freedom9339 replied:
Fixed.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,122 @@
+    selector: 'move-controller-service',
+    standalone: true,
+    templateUrl: './move-controller-service.component.html',
+    imports: [
```
A number of the `imports` below are no longer used. Once they are removed from here, please ensure the the corresponding `import` above is removed.

Freedom9339 replied:
Removed unused imports.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/flow/flow.selectors.ts:
```diff
@@ -47,6 +47,11 @@ export const selectCurrentParameterContext = createSelector(
+    selectFlowState,
+    (state: FlowState) => state.flow.processGroupFlow
+);
+
```
Removed.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,122 @@
+import {
+    ControllerServiceEntity,
+    ControllerServiceReferencingComponent,
+    ControllerServiceReferencingComponentEntity
```
Removed.

Freedom9339 commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/controller-service/move-controller-service/move-controller-service.component.ts:
```diff
@@ -0,0 +1,122 @@
+import { NifiSpinnerDirective } from 'apps/nifi/src/app/ui/common/spinner/nifi-spinner.directive';
+import { MoveControllerServiceDialogRequestSuccess } from '../../../state/controller-services';
+import { moveControllerService } from '../../../state/controller-services/controller-services.actions';
+import { BreadcrumbEntity } from '../../../state/shared';
```
Removed.

Freedom9339 commented:
@mcgilman Sorry for the delay. I addressed the issues you mentioned and added more user validation to not expose any unauthorized components. I've tested every scenario I could think of. Please let me know if there is one I missed.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/ui/controller-service/move-controller-service/move-controller-service.component.html:
```diff
@@ -0,0 +1,87 @@
+                                    [tooltipInputData]="option.description"
+                                    [delayClose]="false">
+                                    <mat-option [value]="option.value" [disabled]="option.disabled">
+                                        <div
+                                            *ngIf="option.description != undefined; else valid"
+                                            class="pointer fa fa-warning has-errors caution-color"
+                                            style="padding: 3px"></div>
```
I'm not sure we should be using the warning icon here. When we use this, the user can hover to reveal the issue. However, here the icon cannot be interacted with and instead the warning is revealed as a tooltip on the menu item.

Freedom9339 replied:
I've removed the warning icon.

mcgilman commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ControllerServiceResource.java:
```diff
@@ -664,6 +668,206 @@ public Response updateControllerService(
+            authorized = authorized && authorizableProcessGroupCurrent.getAuthorizable().isAuthorized(authorizer, RequestAction.WRITE, user);
+
+            if (authorized) {
+                if (authorizableProcessGroupCurrent.getProcessGroup().getParent() != null) {
+                    final ProcessGroupAuthorizable authorizableProcessGroupParent = lookup.getProcessGroup(authorizableProcessGroupCurrent.getProcessGroup().getParent().getIdentifier());
+                    if (authorizableProcessGroupParent.getAuthorizable().isAuthorized(authorizer, RequestAction.READ, user)
+                            && authorizableProcessGroupParent.getAuthorizable().isAuthorized(authorizer, RequestAction.WRITE, user)) {
+                        options.add(generateProcessGroupOption(controllerServiceDTO, authorizableProcessGroupParent, lookup, user));
+                    }
+                }
+
+                authorizableProcessGroupCurrent.getProcessGroup().getProcessGroups().forEach(processGroup -> {
+                    final ProcessGroupAuthorizable authorizableProcessGroup = lookup.getProcessGroup(processGroup.getIdentifier());
+                    if (authorizableProcessGroup.getAuthorizable().isAuthorized(authorizer, RequestAction.READ, user)
+                            && authorizableProcessGroup.getAuthorizable().isAuthorized(authorizer, RequestAction.WRITE, user)) {
+                        options.add(generateProcessGroupOption(controllerServiceDTO, authorizableProcessGroup, lookup, user));
+                    }
+                });
```
Both parent and child options are added to the same collection. Do you think it would be helpful to differentiate between the parent Process Group from the child Process Groups in the listing shown to the user. I'm not suggesting we have to do this, just looking for feedback if you think it may be helpful to the user especially when dealing with a large flow.

The options collection is already ordered and should position the parent Process Group first but it'll only be present when the user has permissions.

mcgilman commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ControllerServiceResource.java:
```diff
@@ -664,6 +668,206 @@ public Response updateControllerService(
+        return generateOkResponse(options).build();
+    }
+
+    private ProcessGroupOptionEntity generateProcessGroupOption(ControllerServiceDTO controllerServiceDTO, ProcessGroupAuthorizable processGroup, AuthorizableLookup lookup, NiFiUser user) {
```
This sort of logic is typically within the service facade. The Resource tier handles request validation, authorization, and request replication. It would defer to the service facade to handle the actual logic of the request.

Freedom9339 replied:
I've moved all the logic to the service facade.

mcgilman commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ControllerServiceResource.java:
```diff
@@ -664,6 +668,206 @@ public Response updateControllerService(
+                    @SecurityRequirement(name = "Read - /flow")
+            }
+    )
+    public Response getProcessGroupOptions(
```
This endpoint needs to be replicated across the cluster. NiFi supports an extensible authorization model. That considered, for endpoints like we the call would be replicated to each node in the cluster. When in the response merger, we need to essentially return the intersection of options that are approved and in common for all nodes in the cluster.

Freedom9339 replied:
Added endpoint replication for clusters.

exceptionfactory replied:
The method implementation still appears to be limited to the current node, as opposed to replicating the request. In addition, I don't see the response merger implemented.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -658,6 +659,86 @@ export class ControllerServicesEffects {
+                        })
+                    ),
+                    catchError((errorResponse: HttpErrorResponse) =>
+                        of(ErrorActions.snackBarError({ error: this.errorHelper.getErrorString(errorResponse) }))
```
Since this action happens with the dialog open and the dialog remains open, we should be using a banner error here. This will show the error in a banner within the dialog.

Freedom9339 replied:
I've changed this to use the banner instead of a toast.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -658,6 +659,86 @@ export class ControllerServicesEffects {
+                        }),
+                        catchError((errorResponse: HttpErrorResponse) => {
+                            this.dialog.closeAll();
+                            return of(
+                                ErrorActions.snackBarError({ error: this.errorHelper.getErrorString(errorResponse) })
+                            );
```
The effects for this action does not dispatch so this `return` has no effect. You'll want to replace this with `this.store.dispatch(...)`

Freedom9339 replied:
Changed this to tap and used store.dispatch. Error show up correctly now.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.reducer.ts:
```diff
@@ -99,6 +100,17 @@ export const controllerServicesReducer = createReducer(
             draftState.saving = false;
         });
     }),
+    on(moveControllerServiceSuccess, (state, { response }) => {
```
`moveControllerService` should be added to the reducer above to ensure that the `saving` flag is set to `true`.

Freedom9339 replied:
I've added it to the reducers.

mcgilman reviewed (changes requested):
Thanks for the updates and the mention @Freedom9339! The front end changes are starting to look good. I've noted a couple minor things below. I've also noted a few things on the back end code but we'll want to get a closer look from @markap14 or @exceptionfactory before we sign off.

Freedom9339 commented:
@mcgilman Thank you for the review. I've made the changed requested.

Freedom9339 commented:
@mcgilman Any new updates on this?

mcgilman commented:
@Freedom9339 In my last review I mentioned the UI portions were looking good but I will take a peek at your latest updates. I also provided some initial feedback on some of the back end changes but we'll want some additional eyes on those from someone more familiar with those portions of the code base.

mcgilman commented on nifi-frontend/src/main/frontend/apps/nifi/src/app/pages/flow-designer/state/controller-services/controller-services.effects.ts:
```diff
@@ -658,6 +659,98 @@ export class ControllerServicesEffects {
+                                data: {
+                                    controllerService: request.controllerService,
+                                    options: response
+                                }
+                            });
+
+                            moveDialogReference.componentInstance.goToReferencingComponent = (
+                                component: ControllerServiceReferencingComponent
+                            ) => {
+                                const route: string[] = this.getRouteForReference(component);
+                                this.router.navigate(route);
+                            };
+
+                            moveDialogReference.afterClosed().subscribe((response) => {
+                                if (response != 'ROUTED') {
+                                    this.store.dispatch(
+                                        ControllerServicesActions.loadControllerServices({
+                                            request: {
+                                                processGroupId: request.controllerService.parentGroupId!
+                                            }
+                                        })
+                                    );
+                                }
+                            });
+                        }),
+                        tap({
+                            error: (errorResponse: HttpErrorResponse) => {
+                                this.dialog.closeAll();
+                                this.store.dispatch(
+                                    ErrorActions.snackBarError({
+                                        error: this.errorHelper.getErrorString(errorResponse)
+                                    })
+                                );
+                            }
+                        })
+                    )
+                )
+            ),
+        { dispatch: false }
+    );
```
```suggestion
    openMoveControllerServiceDialog$ = createEffect(
        () =>
            this.actions$.pipe(
                ofType(ControllerServicesActions.openMoveControllerServiceDialog),
                map((action) => action.request),
                switchMap((request) =>
                    from(this.controllerServiceService.getMoveOptions(request.controllerService.id)).pipe(
                        map((options: SelectOption[]) => {
                            return { request, options };
                        }),
                        tap({
                            error: (errorResponse: HttpErrorResponse) => {
                                this.store.dispatch(
                                    ErrorActions.snackBarError({
                                        error: this.errorHelper.getErrorString(errorResponse)
                                    })
                                );
                            }
                        })
                    )
                ),
                tap(({ request, options }) => {
                    const moveDialogReference = this.dialog.open(MoveControllerService, {
                        ...LARGE_DIALOG,
                        data: {
                            controllerService: request.controllerService,
                            options
                        }
                    });

                    moveDialogReference.componentInstance.goToReferencingComponent = (
                        component: ControllerServiceReferencingComponent
                    ) => {
                        const route: string[] = this.getRouteForReference(component);
                        this.router.navigate(route);
                    };

                    moveDialogReference.afterClosed().subscribe((response) => {
                        if (response != 'ROUTED') {
                            this.store.dispatch(
                                ControllerServicesActions.loadControllerServices({
                                    request: {
                                        processGroupId: request.controllerService.parentGroupId!
                                    }
                                })
                            );
                        }
                    });
                })
            ),
        { dispatch: false }
    );
```

Here's a suggestion for the effect for querying for the available options prior to opening the dialog.

Freedom9339 replied:
Applied this suggestion.

exceptionfactory reviewed (changes requested):
Thanks for the substantive effort on this feature @Freedom9339, and for the significant review work @mcgilman.

Unfortunately, I think the length of time this pull request has remained open, and the remaining issues around the service implementation, raise a concern about this going forward. As evidenced by the amount of changes required thus far, this feature presents a number of complexities related to request replication and permission checking that must be implemented correctly. Part of the challenge relates to the current complexity surrounding fine-grained permissions.

Although this provides helpful background, it seems like it may require a project maintainer to bring it to a state that is ready for merging. Again, this is primarily due to the significant complexity around permissions and clustering, requiring direct development as opposed a more standard pull request feedback cycle. That may mean this is put on hold for a time until there are sufficient cycles to bring it to completion.

Freedom9339 commented:
@exceptionfactory I've added the replicating of the move-options end point as well as merging the responses. Please let me know if this is being placed on hold or if there is anything else I need to do.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ControllerServiceResource.java:
```diff
@@ -664,6 +668,155 @@ public Response updateControllerService(
+    @PUT
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @Path("{id}/move")
```
Although the REST API is not completely consistent, the general pattern is to have the path as a resource, which would change the approach to have the Process Group in the path.

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/groups/StandardProcessGroup.java:
```diff
@@ -2580,7 +2580,9 @@ public void removeControllerService(final ControllerServiceNode service) {
             }
 
-            service.verifyCanDelete();
+            if (!service.isMoving()) {
```
Is there a reason for not checking moving status inside the `verifyCanDelete` method?

exceptionfactory commented on nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/service/StandardControllerServiceNode.java:
```diff
@@ -134,6 +135,7 @@ public StandardControllerServiceNode(final LoggableComponent<ControllerService>
         super(id, validationContextFactory, serviceProvider, componentType, componentCanonicalClass, reloadComponent, extensionManager, validationTrigger, isExtensionMissing);
         this.serviceProvider = serviceProvider;
         this.active = new AtomicBoolean();
+        this.moving = new AtomicBoolean(false);
```
Adding this temporary state to the Node object raises a number of questions about the implementation, as it does not seem to fit with the other stateful properties.

exceptionfactory reviewed (changes requested):
@Freedom9339 I raised a few questions and noted several concerns on the REST API and service implementation. As mentioned, this is a complex feature, so I appreciate the effort. However, it is complex enough that it appears to require direct maintainer involvement to move forward. The work here is very helpful background, I'm not in a position to work on it right now, so it seems best to leave and restart down the road.

joewitt commented:
Given how long this has been in process and the recent rounds of feedback I think we should close this.

The effort that has gone into this contribution is very impressive as is the depth of reviews from very experienced members of the community.  But this is too big/too complex and needs to be very consistent to land.
